cli: Example_demo failed

[(cli).Example_demo failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2453878&tab=buildLog) on [master@06944cb1996ed9ccfa036118b810e46673fe0ff4](https://github.com/cockroachdb/cockroach/commits/06944cb1996ed9ccfa036118b810e46673fe0ff4):

Fatal error:
```
panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer [recovered]
	panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer

```

Stack:
```
goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc004a85ca0, 0xc0044a8700, 0xef, 0x4925bcd2, 0x410b3e0, 0xc0055ee750, 0x432b54)
	/usr/local/go/src/testing/example.go:89 +0x648
testing.runExample.func2(0xbfe54d6cdc75a887, 0x1d3cea0884, 0x7678880, 0xc000466528, 0xc000126008, 0xc003a4e900, 0xc004a85ca0, 0xc004a85cd0)
	/usr/local/go/src/testing/run_example.go:58 +0x10d
panic(0x410b3e0, 0xc0055ee750)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/cli.(*transientCluster).start(0xc00755cc80, 0x53d0200, 0xc000124010, 0x734f820, 0x0, 0x0, 0xc004a852c0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo_cluster.go:162 +0x11ce
github.com/cockroachdb/cockroach/pkg/cli.runDemo(0x734f820, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:279 +0x38c
github.com/cockroachdb/cockroach/pkg/cli.glob..func8(0x734f820, 0xc0052cf9e0, 0x0, 0x3, 0xc0038b6cf0, 0xd)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:52 +0x33
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x734f820, 0xc0052cf9e0, 0x0, 0x3, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:193 +0x79
github.com/spf13/cobra.(*Command).execute(0x734f820, 0xc0052cf8f0, 0x3, 0x3, 0x734f820, 0xc0052cf8f0)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x73469a0, 0xc000466528, 0xc004a858d0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914 +0x30b
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:226
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs.func1(0xc0034ee6c0, 0x4, 0x4, 0x0, 0xc00105c4b0, 0x44, 0x4d7e578, 0xc000126690, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:359 +0x2ab
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs(0x0, 0xc00105c4b0, 0x44, 0x4d7e578, 0xc000126690, 0x0, 0x0, 0x0, 0xc0034ee6c0, 0x4, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:360 +0x85
github.com/cockroachdb/cockroach/pkg/cli.Example_demo()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:448 +0x7f3
testing.runExample(0x46d3356, 0xc, 0x4d7dfc0, 0x4c4cfcc, 0x45e, 0x0, 0x0)
	/usr/local/go/src/testing/run_example.go:62 +0x209
testing.runExamples(0xc004a85e48, 0x7358fe0, 0x18, 0x18, 0xbfe54fa579c21700)
	/usr/local/go/src/testing/example.go:44 +0x1af
testing.(*M).Run(0xc000d1bd00, 0x0)
	/usr/local/go/src/testing/testing.go:1346 +0x273
github.com/cockroachdb/cockroach/pkg/cli_test.TestMain(0xc000d1bd00)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/main_test.go:34 +0x8e
main.main()
	_testmain.go:235 +0x165

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   Example_demo
[demo -e show database]
W201118 06:24:52.199831 1 server/status/runtime.go:311  [n?] could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
E201118 06:24:52.207400 1 cli/error.go:398  listen tcp 127.0.0.1:8080: bind: address already in use
[demo -e show database --empty]
--- FAIL: Example_demo (1.23s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=Example_demo PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2AExample_demo.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(cli).Example_demo failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2453879&tab=buildLog) on [master@06944cb1996ed9ccfa036118b810e46673fe0ff4](https://github.com/cockroachdb/cockroach/commits/06944cb1996ed9ccfa036118b810e46673fe0ff4):

Fatal error:
```
panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer [recovered]
	panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer

```

Stack:
```
goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc02dd3fb20, 0xc03dc01a40, 0xef, 0x14d75584c, 0x61c2200, 0xc02a1bb230, 0xc02dd3dd18)
	/usr/local/go/src/testing/example.go:89 +0x731
testing.runExample.func2(0xbfe54e358136345b, 0xc3b26dd45f, 0xa695b00, 0xc02def8c18, 0xc000126008, 0xc01e1be480, 0xc02dd3fb20, 0xc02dd3fb50)
	/usr/local/go/src/testing/run_example.go:58 +0x148
panic(0x61c2200, 0xc02a1bb230)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/cli.(*transientCluster).start(0xc01f3dc900, 0x81d0d60, 0xc000124010, 0xa366880, 0x0, 0x0, 0xc02dd3ed50, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo_cluster.go:162 +0x1c59
github.com/cockroachdb/cockroach/pkg/cli.runDemo(0xa366880, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:279 +0x416
github.com/cockroachdb/cockroach/pkg/cli.glob..func8(0xa366880, 0xc025419f20, 0x0, 0x3, 0x4769c5, 0x5d6f3a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:52 +0x45
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0xa366880, 0xc025419f20, 0x0, 0x3, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:193 +0x9a
github.com/spf13/cobra.(*Command).execute(0xa366880, 0xc025419e90, 0x3, 0x3, 0xa366880, 0xc025419e90)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826 +0x536
github.com/spf13/cobra.(*Command).ExecuteC(0xa35da00, 0xc02def8c18, 0xc02dd3f5b8, 0x1)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914 +0x432
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:226
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs.func1(0xc00d72d480, 0x4, 0x4, 0x0, 0xc018ff3630, 0x44, 0x6e3da18, 0xc000126690, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:359 +0x365
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs(0x0, 0xc018ff3630, 0x44, 0x6e3da18, 0xc000126690, 0x0, 0x0, 0x0, 0xc00d72d480, 0x4, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:360 +0xe5
github.com/cockroachdb/cockroach/pkg/cli.Example_demo()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:448 +0x121e
testing.runExample(0x6791e04, 0xc, 0x6e3d460, 0x6d0c713, 0x45e, 0x0, 0x0)
	/usr/local/go/src/testing/run_example.go:62 +0x2b6
testing.runExamples(0xc02dd3fdf0, 0xa36fbc0, 0x17, 0x17, 0xbfe54fbb703fa5f4)
	/usr/local/go/src/testing/example.go:44 +0x229
testing.(*M).Run(0xc0005b6200, 0x0)
	/usr/local/go/src/testing/testing.go:1346 +0x465
github.com/cockroachdb/cockroach/pkg/cli_test.TestMain(0xc0005b6200)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/main_test.go:34 +0xb0
main.main()
	_testmain.go:233 +0x272

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   Example_demo
[demo -e show database]
W201118 06:38:17.488140 1 server/status/runtime.go:311  [n?] could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
E201118 06:38:17.533167 1 cli/error.go:398  listen tcp 127.0.0.1:8080: bind: address already in use
[demo -e show database --empty]
--- FAIL: Example_demo (5.59s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=Example_demo PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2AExample_demo.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(cli).Example_demo failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2457460&tab=buildLog) on [master@8ff64f5d897a9c7b39ca437193a11a7300cecbc1](https://github.com/cockroachdb/cockroach/commits/8ff64f5d897a9c7b39ca437193a11a7300cecbc1):

Fatal error:
```
panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer [recovered]
	panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer

```

Stack:
```
goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc004e87ca0, 0xc002aad180, 0xef, 0x4e4eab83, 0x4115000, 0xc00465d1a0, 0x432b54)
	/usr/local/go/src/testing/example.go:89 +0x648
testing.runExample.func2(0xbfe5a1d1087fee2e, 0x1cce782a46, 0x767e920, 0xc000019848, 0xc000126008, 0xc0033dbb60, 0xc004e87ca0, 0xc004e87cd0)
	/usr/local/go/src/testing/run_example.go:58 +0x10d
panic(0x4115000, 0xc00465d1a0)
	/usr/local/go/src/runtime/panic.go:969 +0x1b9
github.com/cockroachdb/cockroach/pkg/cli.(*transientCluster).start(0xc005595880, 0x53d5500, 0xc000124010, 0x73558c0, 0x0, 0x0, 0xc004e872c0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo_cluster.go:162 +0x11ce
github.com/cockroachdb/cockroach/pkg/cli.runDemo(0x73558c0, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:279 +0x38c
github.com/cockroachdb/cockroach/pkg/cli.glob..func8(0x73558c0, 0xc0038bedb0, 0x0, 0x3, 0xc002dca8d0, 0xd)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:52 +0x33
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0x73558c0, 0xc0038bedb0, 0x0, 0x3, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:193 +0x79
github.com/spf13/cobra.(*Command).execute(0x73558c0, 0xc0038bed80, 0x3, 0x3, 0x73558c0, 0xc0038bed80)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826 +0x47c
github.com/spf13/cobra.(*Command).ExecuteC(0x734ca40, 0xc000019848, 0xc004e878d0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914 +0x30b
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:226
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs.func1(0xc00438ec40, 0x4, 0x4, 0x0, 0xc00491d810, 0x44, 0x4d89530, 0xc000126698, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:359 +0x2ab
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs(0x0, 0xc00491d810, 0x44, 0x4d89530, 0xc000126698, 0x0, 0x0, 0x0, 0xc00438ec40, 0x4, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:360 +0x85
github.com/cockroachdb/cockroach/pkg/cli.Example_demo()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:448 +0x7f3
testing.runExample(0x46ddbfc, 0xc, 0x4d88f78, 0x4c57f11, 0x45e, 0x0, 0x0)
	/usr/local/go/src/testing/run_example.go:62 +0x209
testing.runExamples(0xc004e87e48, 0x735f080, 0x18, 0x18, 0xbfe5a40a1d1b1436)
	/usr/local/go/src/testing/example.go:44 +0x1af
testing.(*M).Run(0xc000f86300, 0x0)
	/usr/local/go/src/testing/testing.go:1346 +0x273
github.com/cockroachdb/cockroach/pkg/cli_test.TestMain(0xc000f86300)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/main_test.go:34 +0x8e
main.main()
	_testmain.go:235 +0x165

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   Example_demo
[demo -e show database]
W201119 06:25:08.762083 1 server/status/runtime.go:311  [n?] could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
E201119 06:25:08.769995 1 cli/error.go:398  listen tcp 127.0.0.1:8080: bind: address already in use
[demo -e show database --empty]
--- FAIL: Example_demo (1.31s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=Example_demo PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2AExample_demo.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(cli).Example_demo failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2457461&tab=buildLog) on [master@8ff64f5d897a9c7b39ca437193a11a7300cecbc1](https://github.com/cockroachdb/cockroach/commits/8ff64f5d897a9c7b39ca437193a11a7300cecbc1):

Fatal error:
```
panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer [recovered]
	panic: interface conversion: interface {} is *withstack.withStack, not *server.TestServer

```

Stack:
```
goroutine 1 [running]:
testing.(*InternalExample).processRunResult(0xc02cc67b20, 0xc0267d3dc0, 0xef, 0xfa4b9f3a, 0x6199500, 0xc01b571980, 0xc02cc65d18)
	/usr/local/go/src/testing/example.go:89 +0x731
testing.runExample.func2(0xbfe5a29c4a789fa9, 0xc0f3179e47, 0xa676b80, 0xc044555e10, 0xc000010018, 0xc00fe6aea0, 0xc02cc67b20, 0xc02cc67b50)
	/usr/local/go/src/testing/run_example.go:58 +0x148
panic(0x6199500, 0xc01b571980)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
github.com/cockroachdb/cockroach/pkg/cli.(*transientCluster).start(0xc013b21d80, 0x81b46a0, 0xc0000740f8, 0xa347920, 0x0, 0x0, 0xc02cc66d50, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo_cluster.go:162 +0x1c59
github.com/cockroachdb/cockroach/pkg/cli.runDemo(0xa347920, 0x0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:279 +0x416
github.com/cockroachdb/cockroach/pkg/cli.glob..func8(0xa347920, 0xc00bcf7c50, 0x0, 0x3, 0x4769c5, 0x5d45720)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/demo.go:52 +0x45
github.com/cockroachdb/cockroach/pkg/cli.MaybeDecorateGRPCError.func1(0xa347920, 0xc00bcf7c50, 0x0, 0x3, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/error.go:193 +0x9a
github.com/spf13/cobra.(*Command).execute(0xa347920, 0xc00bcf7b30, 0x3, 0x3, 0xa347920, 0xc00bcf7b30)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:826 +0x536
github.com/spf13/cobra.(*Command).ExecuteC(0xa33eaa0, 0xc044555e10, 0xc02cc675b8, 0x1)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914 +0x432
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
github.com/cockroachdb/cockroach/pkg/cli.Run(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli.go:226
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs.func1(0xc00cbc9a00, 0x4, 0x4, 0x0, 0xc01e101ef0, 0x44, 0x6e16090, 0xc0000106a0, 0x0, 0x0, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:359 +0x365
github.com/cockroachdb/cockroach/pkg/cli.cliTest.RunWithArgs(0x0, 0xc01e101ef0, 0x44, 0x6e16090, 0xc0000106a0, 0x0, 0x0, 0x0, 0xc00cbc9a00, 0x4, ...)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:360 +0xe5
github.com/cockroachdb/cockroach/pkg/cli.Example_demo()
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/cli_test.go:448 +0x121e
testing.runExample(0x6769d6a, 0xc, 0x6e15ad8, 0x6ce4d18, 0x45e, 0x0, 0x0)
	/usr/local/go/src/testing/run_example.go:62 +0x2b6
testing.runExamples(0xc02cc67df0, 0xa350c60, 0x17, 0x17, 0xbfe5a4252d85f9ac)
	/usr/local/go/src/testing/example.go:44 +0x229
testing.(*M).Run(0xc000424600, 0x0)
	/usr/local/go/src/testing/testing.go:1346 +0x465
github.com/cockroachdb/cockroach/pkg/cli_test.TestMain(0xc000424600)
	/go/src/github.com/cockroachdb/cockroach/pkg/cli/main_test.go:34 +0xb0
main.main()
	_testmain.go:233 +0x272

```

<details><summary>Log preceding fatal error</summary><p>

```
=== RUN   Example_demo
[demo -e show database]
W201119 06:38:43.960302 1 server/status/runtime.go:311  [n?] could not parse build timestamp: parsing time "" as "2006/01/02 15:04:05": cannot parse "" as "2006"
E201119 06:38:44.012766 1 cli/error.go:398  listen tcp 127.0.0.1:8080: bind: address already in use
[demo -e show database --empty]
--- FAIL: Example_demo (4.20s)

```

</p></details>

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=Example_demo PKG=./pkg/cli TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2AExample_demo.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/cli/cli_test.go
+++ b/pkg/cli/cli_test.go
@@ -419,6 +419,9 @@ func Example_demo() {
 	c := newCLITest(cliTestParams{noServer: true})
 	defer c.cleanup()
 
+	defer func(b bool) { testingForceRandomizeDemoPorts = b }(testingForceRandomizeDemoPorts)
+	testingForceRandomizeDemoPorts = true
+
 	testData := [][]string{
 		{`demo`, `-e`, `show database`},
 		{`demo`, `-e`, `show database`, `--empty`},

--- a/pkg/cli/demo_cluster.go
+++ b/pkg/cli/demo_cluster.go
@@ -326,15 +326,10 @@ func testServerArgsForTransientCluster(
 	storeSpec := base.DefaultTestStoreSpec
 	storeSpec.StickyInMemoryEngineID = fmt.Sprintf("demo-node%d", nodeID)
 
-	sqlPort := sqlBasePort + int(nodeID) - 1
-	httpPort := httpBasePort + int(nodeID) - 1
-
 	args := base.TestServerArgs{
 		SocketFile:              sock.filename(),
 		PartOfCluster:           true,
 		Stopper:                 stop.NewStopper(),
-		SQLAddr:                 fmt.Sprintf(":%d", sqlPort),
-		HTTPAddr:                fmt.Sprintf(":%d", httpPort),
 		JoinAddr:                joinAddr,
 		DisableTLSForHTTP:       true,
 		StoreSpecs:              []base.StoreSpec{storeSpec},
@@ -344,6 +339,15 @@ func testServerArgsForTransientCluster(
 		// This disables the tenant server. We could enable it but would have to
 		// generate the suitable certs at the caller who wishes to do so.
 		TenantAddr: new(string),
+	}
+
+	if !testingForceRandomizeDemoPorts {
+		// Unit tests can be run with multiple processes side-by-side with
+		// `make stress`. This is bound to not work with fixed ports.
+		sqlPort := sqlBasePort + int(nodeID) - 1
+		httpPort := httpBasePort + int(nodeID) - 1
+		args.SQLAddr = fmt.Sprintf(":%d", sqlPort)
+		args.HTTPAddr = fmt.Sprintf(":%d", httpPort)
 	}
 
 	if demoCtx.localities != nil {
@@ -358,6 +362,10 @@ func testServerArgsForTransientCluster(
 
 	return args
 }
+
+// testingForceRandomizeDemoPorts disables the fixed port allocation
+// for demo clusters, for use in tests.
+var testingForceRandomizeDemoPorts bool
 
 func (c *transientCluster) cleanup(ctx context.Context) {
 	if c.stopper != nil {

--- a/pkg/cli/demo_cluster.go
+++ b/pkg/cli/demo_cluster.go
@@ -159,7 +159,12 @@ func (c *transientCluster) start(
 			}
 		}
 
-		serv := serverFactory.New(args).(*server.TestServer)
+		s, err := serverFactory.New(args)
+		if err != nil {
+			return err
+		}
+		serv := s.(*server.TestServer)
+		c.stopper.AddCloser(stop.CloserFn(serv.Stop))
 		if i == 0 {
 			c.s = serv
 			// The first node connects its Settings instance to the `log`
@@ -204,7 +209,6 @@ func (c *transientCluster) start(
 			errCh <- nil
 		}
 
-		c.stopper.AddCloser(stop.CloserFn(serv.Stop))
 		// Ensure we close all sticky stores we've created.
 		for _, store := range args.StoreSpecs {
 			if store.StickyInMemoryEngineID != "" {
@@ -491,7 +495,11 @@ func (c *transientCluster) RestartNode(nodeID roachpb.NodeID) error {
 	// TODO(#42243): re-compute the latency mapping.
 	args := testServerArgsForTransientCluster(c.sockForServer(nodeID), nodeID, c.s.ServingRPCAddr(), c.demoDir,
 		c.sqlFirstPort, c.httpFirstPort)
-	serv := server.TestServerFactory.New(args).(*server.TestServer)
+	s, err := server.TestServerFactory.New(args)
+	if err != nil {
+		return err
+	}
+	serv := s.(*server.TestServer)
 
 	// We want to only return after the server is ready.
 	readyCh := make(chan struct{})

--- a/pkg/cli/demo_locality_test.go
+++ b/pkg/cli/demo_locality_test.go
@@ -21,6 +21,9 @@ func Example_demo_locality() {
 	c := newCLITest(cliTestParams{noServer: true})
 	defer c.cleanup()
 
+	defer func(b bool) { testingForceRandomizeDemoPorts = b }(testingForceRandomizeDemoPorts)
+	testingForceRandomizeDemoPorts = true
+
 	testData := [][]string{
 		{`demo`, `--nodes`, `3`, `-e`, `select node_id, locality from crdb_internal.gossip_nodes order by node_id`},
 		{`demo`, `--nodes`, `9`, `-e`, `select node_id, locality from crdb_internal.gossip_nodes order by node_id`},

--- a/pkg/server/settings_cache_test.go
+++ b/pkg/server/settings_cache_test.go
@@ -86,9 +86,12 @@ func TestCachedSettingsServerRestart(t *testing.T) {
 	})
 	testServer.Stopper().Stop(context.Background())
 
-	srv, ok := serverutils.NewServer(serverArgs).(*TestServer)
+	ts, err := serverutils.NewServer(serverArgs)
+	if err != nil {
+		t.Fatal(err)
+	}
+	srv := ts.(*TestServer)
 	defer srv.Stopper().Stop(context.Background())
-	require.True(t, ok)
 
 	s := srv.Server
 	var initServer *initServer

--- a/pkg/server/testserver.go
+++ b/pkg/server/testserver.go
@@ -1291,7 +1291,7 @@ type testServerFactoryImpl struct{}
 var TestServerFactory = testServerFactoryImpl{}
 
 // New is part of TestServerFactory interface.
-func (testServerFactoryImpl) New(params base.TestServerArgs) interface{} {
+func (testServerFactoryImpl) New(params base.TestServerArgs) (interface{}, error) {
 	cfg := makeTestConfigFromParams(params)
 	ts := &TestServer{Cfg: &cfg, params: params}
 
@@ -1306,13 +1306,15 @@ func (testServerFactoryImpl) New(params base.TestServerArgs) interface{} {
 	// Needs to be called before NewServer to ensure resolvers are initialized.
 	ctx := context.Background()
 	if err := ts.Cfg.InitNode(ctx); err != nil {
-		return err
+		params.Stopper.Stop(ctx)
+		return nil, err
 	}
 
 	var err error
 	ts.Server, err = NewServer(*ts.Cfg, params.Stopper)
 	if err != nil {
-		return err
+		params.Stopper.Stop(ctx)
+		return nil, err
 	}
 
 	// Create a breaker which never trips and never backs off to avoid
@@ -1326,5 +1328,5 @@ func (testServerFactoryImpl) New(params base.TestServerArgs) interface{} {
 	// Our context must be shared with our server.
 	ts.Cfg = &ts.Server.cfg
 
-	return ts
+	return ts, nil
 }

--- a/pkg/testutils/reduce/reducesql/reducesql_test.go
+++ b/pkg/testutils/reduce/reducesql/reducesql_test.go
@@ -41,11 +41,15 @@ func isInterestingSQL(contains string) reduce.InterestingFn {
 		args := base.TestServerArgs{
 			Insecure: true,
 		}
-		serv := server.TestServerFactory.New(args).(*server.TestServer)
+		ts, err := server.TestServerFactory.New(args)
+		if err != nil {
+			panic(err)
+		}
+		serv := ts.(*server.TestServer)
+		defer serv.Stopper().Stop(ctx)
 		if err := serv.Start(); err != nil {
 			panic(err)
 		}
-		defer serv.Stopper().Stop(ctx)
 
 		options := url.Values{}
 		options.Add("sslmode", "disable")

--- a/pkg/testutils/serverutils/BUILD.bazel
+++ b/pkg/testutils/serverutils/BUILD.bazel
@@ -21,5 +21,6 @@ go_library(
         "//pkg/util/protoutil",
         "//pkg/util/stop",
         "//pkg/util/uuid",
+        "//vendor/github.com/cockroachdb/errors",
     ],
 )

--- a/pkg/testutils/serverutils/test_server_shim.go
+++ b/pkg/testutils/serverutils/test_server_shim.go
@@ -36,6 +36,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
 	"github.com/cockroachdb/cockroach/pkg/util/uuid"
+	"github.com/cockroachdb/errors"
 )
 
 // TestServerInterface defines test server functionality that tests need; it is
@@ -225,7 +226,7 @@ type TestServerInterface interface {
 // service.
 type TestServerFactory interface {
 	// New instantiates a test server.
-	New(params base.TestServerArgs) interface{}
+	New(params base.TestServerArgs) (interface{}, error)
 }
 
 var srvFactoryImpl TestServerFactory
@@ -243,7 +244,10 @@ func InitTestServerFactory(impl TestServerFactory) {
 func StartServer(
 	t testing.TB, params base.TestServerArgs,
 ) (TestServerInterface, *gosql.DB, *kv.DB) {
-	server := NewServer(params)
+	server, err := NewServer(params)
+	if err != nil {
+		t.Fatalf("%+v", err)
+	}
 	if err := server.Start(); err != nil {
 		t.Fatalf("%+v", err)
 	}
@@ -252,13 +256,17 @@ func StartServer(
 }
 
 // NewServer creates a test server.
-func NewServer(params base.TestServerArgs) TestServerInterface {
+func NewServer(params base.TestServerArgs) (TestServerInterface, error) {
 	if srvFactoryImpl == nil {
-		panic("TestServerFactory not initialized. One needs to be injected " +
+		return nil, errors.AssertionFailedf("TestServerFactory not initialized. One needs to be injected " +
 			"from the package's TestMain()")
 	}
 
-	return srvFactoryImpl.New(params).(TestServerInterface)
+	srv, err := srvFactoryImpl.New(params)
+	if err != nil {
+		return nil, err
+	}
+	return srv.(TestServerInterface), nil
 }
 
 // OpenDBConnE is like OpenDBConn, but returns an error.
@@ -303,7 +311,10 @@ func OpenDBConn(
 // Generally StartServer() should be used. However this function can be used
 // directly when opening a connection to the server is not desired.
 func StartServerRaw(args base.TestServerArgs) (TestServerInterface, error) {
-	server := NewServer(args)
+	server, err := NewServer(args)
+	if err != nil {
+		return nil, err
+	}
 	if err := server.Start(); err != nil {
 		return nil, err
 	}

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -398,7 +398,11 @@ func (tc *TestCluster) AddServer(serverArgs base.TestServerArgs) (*server.TestSe
 		serverArgs.Addr = serverArgs.Listener.Addr().String()
 	}
 
-	s := serverutils.NewServer(serverArgs).(*server.TestServer)
+	srv, err := serverutils.NewServer(serverArgs)
+	if err != nil {
+		return nil, err
+	}
+	s := srv.(*server.TestServer)
 
 	tc.Servers = append(tc.Servers, s)
 	tc.serverArgs = append(tc.serverArgs, serverArgs)

